File: solutions/869_reordered_power_of_2.cpp
#include "aux.cpp"
#include <algorithm>
#include <bitset>
#include <climits>
#include <iostream>
#include <map>
#include <set>
#include <stdio.h>
#include <unordered_map>
#include <vector>

/**

Starting with a positive integer N, we reorder the digits in any order
(including the original order) such that the leading digit is not zero.

Return true if and only if we can do this in a way such that the resulting
number is a power of 2.



Example 1:

Input: 1
Output: true

Example 2:

Input: 10
Output: false

Example 3:

Input: 16
Output: true

Example 4:

Input: 24
Output: false

Example 5:

Input: 46
Output: true

Note:

    1 <= N <= 10^9
**/
using namespace std;

class Solution {
public:
  bool reorderedPowerOf2(int N) {
    string sn = to_string(N);
    sort(sn.begin(), sn.end());

    int xtm = 1;
    while (to_string(xtm).size() < sn.size())
      xtm *= 2;

    while (to_string(xtm).size() == sn.size()) {
      string sx = to_string(xtm);
      sort(sx.begin(), sx.end());
      if (sx == sn)
        return true;
      xtm *= 2;
    }
    return false;
  }
};

int main() {
  Solution s;
  for (int i = 1; i < 100; i++) {
    vector<int> x{i, s.reorderedPowerOf2(i)};
    if (1 == x.back())
      say(x);
  }
}

File: python_solutions/869.reordered-power-of-2.rb
#
# @lc app=leetcode id=869 lang=ruby
#
# [869] Reordered Power of 2
#
# https://leetcode.com/problems/reordered-power-of-2/description/
#
# algorithms
# Medium (50.35%)
# Total Accepted:    8.7K
# Total Submissions: 17.3K
# Testcase Example:  '1'
#
# Starting with a positive integer N, we reorder the digits in any order
# (including the original order) such that the leading digit is not zero.
#
# Return true if and only if we can do this in a way such that the resulting
# number is a power of 2.
#
#
#
#
#
#
#
# Example 1:
#
#
# Input: 1
# Output: true
#
#
#
# Example 2:
#
#
# Input: 10
# Output: false
#
#
#
# Example 3:
#
#
# Input: 16
# Output: true
#
#
#
# Example 4:
#
#
# Input: 24
# Output: false
#
#
#
# Example 5:
#
#
# Input: 46
# Output: true
#
#
#
#
# Note:
#
#
# 1 <= N <= 10^9
#
#
#
#
#
#
#
#
# @param {Integer} n
# @return {Boolean}
def reordered_power_of2(n)
  arr = n.to_s.chars.map(&:to_i).sort
  maxp = arr.reverse.join.to_i
  x = 1
  while x <= maxp
    return true if x.to_s.chars.map(&:to_i).sort == arr
    x *= 2
  end
  false
end

n = 1
p reordered_power_of2(n)
